java/ql/src/Stubs/Stubs.qll

```diff
@@ -0,0 +1,440 @@
+          ) + ">"
+    else result = ""
+}
+
+private string stubGenericMethodParams(Method m) {
+  if m instanceof GenericMethod
+  then
+    result =
+      " <" +
+        concat(int n, TypeVariable param |
+          param = m.(GenericMethod).getTypeParameter(n)
+        |
+          param.getName(), "," order by n
+        ) + "> "
+  else result = ""
+}
+
+private string stubImplementation(Callable c) {
+  if c.isAbstract()
+  then result = ";"
+  else
+    if c instanceof Constructor or c.getReturnType() instanceof VoidType
+    then result = "{}"
+    else result = "{ return " + stubDefaultValue(c.getReturnType()) + "; }"
+}
+
+private string stubDefaultValue(Type t) {
+  if t instanceof RefType
+  then result = "null"
+  else
+    if t instanceof CharacterType
+    then result = "'0'"
+    else
+      if t instanceof BooleanType
+      then result = "false"
+      else
+        if t instanceof NumericType
+        then result = "0"
+        else result = "<error>"
+}
+
+private string stubParameters(Callable c) {
+  result =
+    concat(int i, Parameter param |
+      param = c.getParameter(i)
+    |
+      stubParameter(param), ", " order by i
+    )
+}
+
+private string stubParameter(Parameter p) {
+  exists(Type t, string suff | result = stubTypeName(t) + suff + " " + p.getName() |
+    if p.isVarargs()
+    then (
+      t = p.getType().(Array).getElementType() and
+      suff = "..."
+    ) else (
+      t = p.getType() and suff = ""
+    )
+  )
+}
+
+private string stubEnumConstants(RefType t) {
+  if t instanceof EnumType
+  then
+    exists(EnumType et | et = t |
+      result =
+        "    " + concat(EnumConstant c | c = et.getAnEnumConstant() | c.getName(), ", ") + ";\n"
+    )
+  else result = ""
+}
+
+// Holds if the member is to be excluded from stubMember
+private predicate excludedMember(Member m) {
+  m instanceof EnumConstant
+  or
+  exists(Method c | m = c |
+    c.getDeclaringType() instanceof EnumType and
+    m.hasName(["values", "valueOf"]) and
+    m.isStatic()
+  )
+}
+
+private string stubMember(Member m) {
+  if excludedMember(m)
+  then result = ""
+  else (
+    exists(Method c | m = c |
+      result =
+        "    " + stubModifiers(c) + stubGenericMethodParams(c) + stubTypeName(c.getReturnType()) +
+          " " + c.getName() + "(" + stubParameters(c) + ")" + stubImplementation(c) + "\n"
+    )
+    or
+    exists(Constructor c | m = c |
+      result =
+        "    " + stubModifiers(m) + c.getName() + "(" + stubParameters(c) + ")" +
+          stubImplementation(c) + "\n"
+    )
+    or
+    exists(Field f | f = m |
+      result =
+        "    " + stubModifiers(m) + stubTypeName(f.getType()) + " " + f.getName() + " = " +
+          stubDefaultValue(f.getType()) + ";\n"
+    )
+    or
+    exists(NestedType nt | nt = m | result = indent(nt.(GeneratedType).getStub()))
+  )
+}
+
+bindingset[s]
+private string indent(string s) { result = "    " + s.replaceAll("\n", "\n    ") + "\n" }
```

Comment:
I'd suggest our Java-generating queries shouldn't attempt formatting, and should instead defer to an external autoformatter